include/revng/RestructureCFG/MarkerCallUtils.h

```diff
@@ -0,0 +1,222 @@
+#pragma once
+
+//
+// Copyright rev.ng Labs Srl. See LICENSE.md for details.
+//
+
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/GlobalValue.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/Support/ModRef.h"
+
+#include "revng/Support/FunctionTags.h"
+#include "revng/Support/IRHelpers.h"
+
+std::string ScopeCloserMarkerCallFunctionName = "scope-closer";
+std::string GotoBlockMarkerCallFunctionName = "goto-block";
+
+inline llvm::Function *
+getMarkerCallFunction(FunctionTags::Tag &MarkerFunctionTag, llvm::Module *M) {
+  llvm::Function *MarkerCallFunction = nullptr;
+
+  // We could early break from this loop but we would loose the ability of
+  // asserting that a single marker function is present
+  for (llvm::Function &F : MarkerFunctionTag.functions(M)) {
+    revng_assert(not MarkerCallFunction);
+    MarkerCallFunction = &F;
+  }
+
+  revng_assert(MarkerCallFunction);
+  return MarkerCallFunction;
+}
+
+inline std::string getMarkerCallFunctionName(llvm::Function *F) {
+  return F->getName().str();
+}
+
+/// This is a helper class used to create the markers in the basic blocks needed
+/// to handle virtual edges, such as the scope closer edges or the goto edges.
+/// Its role should be similar somewhat to the `llvm::IRBuilder` role.
+class MarkerCallBuilder {
+private:
+  llvm::BasicBlock *BB = nullptr;
+  llvm::Function *MarkerCallF = nullptr;
+
+public:
+  MarkerCallBuilder(FunctionTags::Tag &MarkerFunctionTag, llvm::Function *F) {
```

Comment:
Mmh so you're making a builder for each kind of markers.

I think we should only make a single builder for both markers.

You can then have better APIs:
- `makeGoto(BasicBlock *)`
- `addScopeCloser(BasicBlock *Source, BasicBlock *Target)`

Reply:
And then `insertMarkerCall` which is very unspecific can just go away.

Reply:
I added the `makeGoto` and `addScopeCloser` method here https://github.com/revng/revng/blob/57e9b483c79b868ba12fde41644a3ee0aba57651/include/revng/RestructureCFG/ScopeGraphUtils.h#L109 and here https://github.com/revng/revng/blob/57e9b483c79b868ba12fde41644a3ee0aba57651/include/revng/RestructureCFG/ScopeGraphUtils.h#L128.

A couple of observations due to the refactoring you suggested:
- Before the changes the builder was meant to be used as follows: You create the builder, either for creating a `scope_closer` or `goto_block` annotation, so you could create them singularly. Now the builder requires (or creates on demand) both the `scope_closer` and `goto_block` propotypes upon creation, and then you can create the annotations with the `makeGoto` and `addScopeCloser` methods.
- I agree that the API is more clean in this way, and I prefer it too even if it is more strict (you create both the prototypes even if you just need to create, let's say, a single `goto`), but since we are handling a `ScopeGraph` here, I think this is acceptable in any case.